inference.py

```diff
@@ -345,10 +349,62 @@ def T1_T2(fabric, input_ids, model, text_tokenizer, step):
     model.clear_kv_cache()
     return text_tokenizer.decode(torch.tensor(tokenlist)).strip()
 
-    
+
 def load_model(ckpt_dir, device):
+    # Load SNAC model
     snacmodel = SNAC.from_pretrained("hubertsiuzdak/snac_24khz").eval().to(device)
-    whispermodel = whisper.load_model("small").to(device)
+
+    # Load Whisper model for transcription based on the device
+    if device == 'mps':
+        whispermodel = WhisperForConditionalGeneration.from_pretrained("openai/whisper-small").to(device)
+        whisper_processor = WhisperProcessor.from_pretrained("openai/whisper-small")
+
+        def embed_audio(mel):
+            # Convert mel spectrogram to a format suitable for Whisper
+            mel_cpu = mel.cpu().numpy()
+
+            if mel_cpu.ndim > 1:
+                mel_cpu = librosa.to_mono(mel_cpu)
+
+            if mel_cpu.ndim != 1:
+                raise ValueError(f"Audio is not mono! Shape: {mel_cpu.shape}")
+
+            # Process mel with Whisper processor
+            inputs = whisper_processor(mel_cpu, sampling_rate=16000, return_tensors="pt").input_features.to(device)
+
+            # Debugging: Log input feature shapes
+            print(f"Input features shape: {inputs.shape}")
+            print(f"First 5 values of input features: {inputs[0, :, :5]}")
+
+            # Set beam search and max length for transcription
+            generated_ids = whispermodel.generate(
+                inputs,
+                max_length=256,  # Increased length to allow for longer sentences
+                num_beams=5,     # Higher quality output through beam search
+                early_stopping=True
+            )
+
+            # Decode the generated tokens
+            transcription = whisper_processor.decode(generated_ids[0], skip_special_tokens=True)
+
+            # Log the transcription and token IDs
+            print(f"Generated token IDs: {generated_ids[0].tolist()}")
+            print(f"Transcription (for log purposes): {transcription}")
+
+            # Return logits as per original flow
+            start_token = whisper_processor.tokenizer.pad_token_id
+            decoder_input_ids = torch.tensor([[start_token]], device=device)
+
+            outputs = whispermodel(inputs, decoder_input_ids=decoder_input_ids)
+
+            return outputs.logits
```

Comment:
https://github.com/openai/whisper/blob/279133e3107392276dc509148da1f41bfb532c7e/whisper/model.py#L256
please refer to the whisper model, we use embed_audio for encode the audio feature(use only the encoder part of the whisper), and I think it's not the same as your output.logits which is the output of decoder?

Reply:
ah that might be the issue why transcription is wrong, Ok i'll check it